src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseHybridFVM.cpp

```diff
@@ -398,109 +398,124 @@ void SinglePhaseHybridFVM::saveAquiferConvergedState( real64 const & time,
-      arrayView1d< real64 const > const & volume = subRegion.getElementVolume();
+      string const & fluidName = subRegion.getReference< string >( viewKeyStruct::fluidNamesString() );
+      SingleFluidBase const & fluid = getConstitutiveModel< SingleFluidBase >( subRegion, fluidName );
+      defaultViscosity += fluid.defaultViscosity();
+      subRegionCounter++;
 
-      SingleFluidBase const & fluidModel =
-        getConstitutiveModel< SingleFluidBase >( subRegion, subRegion.template getReference< string >( viewKeyStruct::fluidNamesString() ) );
-      arrayView2d< real64 const > const & density_n = fluidModel.density_n();
+      string const & solidName = subRegion.getReference< string >( viewKeyStruct::solidNamesString() );
+      CoupledSolidBase const & solid = getConstitutiveModel< CoupledSolidBase >( subRegion, solidName );
 
-      CoupledSolidBase const & solidModel =
-        SolverBase::getConstitutiveModel< CoupledSolidBase >( subRegion, subRegion.template getReference< string >( viewKeyStruct::solidNamesString() ) );
-      arrayView2d< real64 const > const & porosity_n = solidModel.getPorosity_n();
+      // step 1.1: compute the norm in the subRegion
 
       singlePhaseBaseKernels::
-        ResidualNormKernel::launch< parallelDevicePolicy<> >( localRhs,
-                                                              rankOffset,
-                                                              elemDofNumber,
-                                                              elemGhostRank,
-                                                              volume,
-                                                              density_n,
-                                                              porosity_n,
-                                                              localResidualNorm );
-
-      defaultViscosity += fluidModel.defaultViscosity();
-      subRegionCounter++;
+        ResidualNormKernelFactory::
+        createAndLaunch< parallelDevicePolicy<> >( m_normType,
+                                                   rankOffset,
+                                                   elemDofKey,
+                                                   localRhs,
+                                                   subRegion,
+                                                   fluid,
+                                                   solid,
+                                                   subRegionResidualNorm,
+                                                   subRegionResidualNormalizer );
+
+      // step 1.2: reduction across meshBodies/regions/subRegions
+
+      if( m_normType == solverBaseKernels::NormType::Linf )
+      {
+        if( subRegionResidualNorm[0] > localResidualNorm )
+        {
+          localResidualNorm = subRegionResidualNorm[0];
+        }
+      }
+      else
+      {
+        localResidualNorm += subRegionResidualNorm[0];
+        localResidualNormalizer += subRegionResidualNormalizer[0];
+      }
```

Comment:
This is not really object oriented.
Is there a smooth way not to get this `if`? Some dispatch obviously?

Reply:
I see that you put a couple of comments on the way I did this `if/else`. I think I need help to find a better way to implement this, let's discuss it some time soon.

Reply:
I put this code in helper functions, but it is still not exactly what you asked for. Now the code looks like:
https://github.com/GEOSX/GEOSX/blob/3eada543329ce0bd54db2f4d7adc0f42886f5cb9/src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVM.cpp#L224-L232
It is already slightly better than my first implementation, I think, but the `if/else` is still present.